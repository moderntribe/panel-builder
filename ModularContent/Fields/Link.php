<?php


namespace ModularContent\Fields;
use ModularContent\Panel;


/**
 * Class Link
 *
 * @package ModularContent\Fields
 *
 * A group of fields for creating a link. Includes form controls for
 * the link title, URL, and optionally open in a new window.
 */
class Link extends Field {

	protected $default = [ 'url' => '', 'target' => '', 'label' => '' ];
	protected $layout  = 'compact';


	public function __construct( $args ) {
		$this->check_layout( $args );

		$this->defaults[ 'strings' ] = [
			'placeholder.label' => __( 'Label', 'modular-content' ),
			'placeholder.url'   => __( 'URL', 'modular-content' ),
		];
		$this->defaults['layout'] = $this->layout;
		parent::__construct( $args );
	}

	protected function check_layout( $args ) {
		if ( isset( $args['layout'] ) && $args['layout'] !== 'compact' && $args['layout'] !== 'full' && $args['layout'] !== 'inline' ) {
			throw new \LogicException( 'Layout argument can only be "compact" or "full".' );
		}
	}

	public function get_blueprint() {
		$blueprint           = parent::get_blueprint();
		$blueprint['layout'] = $this->layout;

		return $blueprint;
	}

	/**
	 * Massage submitted data before it's saved.
	 *
	 * @param mixed $data
	 * @return array
	 */
	public function prepare_data_for_save( $data ) {
		return wp_parse_args( $data, [ 'url' => '', 'target' => '', 'label' => '' ] );
	}
} 