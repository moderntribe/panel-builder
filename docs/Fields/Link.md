## Link

A group of fields for creating a link. Includes form controls for the link title, URL, and optionally open in a new window.

### Parameters

||||
|---|---|---|
| `label`       | `string` | The __() translated label for this field       |
| `name`        | `string` | The unique name for this field                 |
| `description` | `string` | The __() translated description for this field |

### Example

```php
$field = new Link( array(
    'label' => __('Link'),
    'name' => 'link',
) );
```