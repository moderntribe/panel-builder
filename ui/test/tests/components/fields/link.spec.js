/* global expect b:true */

import React from 'react';
import { shallow } from 'enzyme';
import Link from 'components/fields/link';
import LinkGroup from 'components/shared/link-group';
import renderer from 'react-test-renderer';


describe('Link field', () => {
	it('renders correctly', () => {
		const tree = renderer
			.create(<Link />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('renders as a <div>', () => {
		const wrapper = shallow(<Link />);
		expect(wrapper.type()).toEqual('div');
	});

	it('has a label', () => {
		const wrapper = shallow(<Link />);
		expect(wrapper.find('legend').length).toEqual(1);
	});

	it('has a LinkGroup', () => {
		const wrapper = shallow(<Link />);
		expect(wrapper.find('fieldset').childAt(1).type()).toEqual(LinkGroup);
	});
});
