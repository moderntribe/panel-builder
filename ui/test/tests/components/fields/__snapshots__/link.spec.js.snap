exports[`Link field renders correctly 1`] = `
<div
  className="field panel-field">
  <fieldset
    className="fieldset">
    <legend
      className="label panel-field-label">
      
    </legend>
    <div>
      <div
        className="inputGeneric panel-group-url">
        <input
          name="url"
          onChange={[Function]}
          placeholder={undefined}
          type="text"
          value="" />
      </div>
      <div
        className="inputGeneric panel-group-label">
        <input
          name="label"
          onChange={[Function]}
          placeholder={undefined}
          type="text"
          value="" />
      </div>
      <div
        className="inputGeneric panel-group-link-target">
        <div
          className="Select Select--single"
          style={undefined}>
          <div
            className="Select-control"
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={undefined}>
            <span
              className="Select-multi-value-wrapper"
              id="react-select-2--value">
              <div
                className="Select-placeholder">
                Select...
              </div>
              <div
                aria-activedescendant="react-select-2--value"
                aria-disabled="false"
                aria-expanded={false}
                aria-label={undefined}
                aria-labelledby={undefined}
                aria-owns=""
                className="Select-input"
                onBlur={[Function]}
                onFocus={[Function]}
                role="combobox"
                style={
                  Object {
                    "border": 0,
                    "display": "inline-block",
                    "width": 1,
                  }
                }
                tabIndex={0} />
            </span>
            <span
              className="Select-arrow-zone"
              onMouseDown={[Function]}>
              <span
                className="Select-arrow"
                onMouseDown={[Function]} />
            </span>
          </div>
        </div>
      </div>
    </div>
  </fieldset>
</div>
`;
